apps/src/blockly/addons/functionEditor.js

```diff
@@ -374,4 +373,31 @@ export default class FunctionEditor {
       blockToUpdate.description = this.block.description;
     }
   }
+
+  // Clear the editor workspace to prepare for a new function definition.
+  clearEditorWorkspace() {
+    // Dispose of all top blocks. We do this manually because we want
+    // to propogate the delete event to the hidden workspace for every block
+    // except the procedure definition. Therefore, we delete all blocks, but for
+    // the procedure definition delete we disable events.
+    while (this.editorWorkspace.getTopBlocks().length) {
+      let isProcedureBlock = false;
+      const topBlock = this.editorWorkspace.getTopBlocks()[0];
+      if (topBlock.id === this.block.id) {
+        isProcedureBlock = true;
+      }
+      if (isProcedureBlock) {
+        Blockly.Events.disable();
+      }
+      topBlock.dispose(false);
+      if (isProcedureBlock) {
+        Blockly.Events.enable();
+      }
+    }
```

Comment:
It took me a while to understand what was going on here with the flipping back and forth of enabling/disabling and checking each element for whether it's the procedure block. What do you think of something like this?

```
clearEditorWorkspace() {
  const topBlocks = this.editorWorkspace.getTopBlocks();
  const orphanedBlocks = topBlocks.filter(blocks => block.id !== this.block.id);
  const procedureDefinitionBlocks = topBlocks.filter(blocks => block.id === this.block.id);

  // Dispose of all non-procedure-definition top blocks (aka orphaned blocks)
  // and propagate the delete event to the hidden workspace
  Blockly.Events.enable();
  orphanedBlocks.forEach(block => block.dispose(false));
  Blockly.Events.disable();

  // Delete the procedure definition block from the editor,
  // but don't delete it from the hidden workspace
  procedureDefinitionBlocks.forEach(block => block.dispose(false));
  this.editorWorkspace.clear();
  Blockly.Events.enable();
}
```

A couple thoughts about this approach:
1. Theoretically, there should only ever be one procedure definition block, which we should probably make a note about, but is handling them all the same the appropriate behavior if there isn't?
2. Would clearing the workspace handle disposing the procedureDefinitionBlock automatically?
3. I would probably use a .reduce() for the first step to avoid doing it in two passes, but I know it's sometimes considered hard to read 😊

Reply:
Your second question maybe resolves are of your other answers: yes clearing the workspace would handle disposing the procedure definition block automatically. So I could do a modified version of the above, where I only find the orphan blocks, dispose of them, then clear the workspace with events disabled to handle the rest. That seems cleaner.

Reply:
updated to what I described above